File: examples/rust/examples/client_chat_advanced.rs
// ekoDB Rust Client - Advanced Chat Features Example
//
// This example demonstrates advanced chat features:
// - Message regeneration
// - Message editing
// - Message deletion
// - Session merging
// - Forgotten messages (exclude from context)

use ekodb_client::{
    Client, MergeSessionsRequest, MergeStrategy, ToggleForgottenRequest, UpdateMessageRequest,
};
use std::env;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("=== ekoDB Advanced Chat Features Example ===\n");

    // Load environment variables
    dotenv::dotenv().ok();

    // Create client
    let base_url = env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string());
    let api_key = env::var("API_BASE_KEY").expect("API_BASE_KEY must be set");

    let client = Client::builder()
        .base_url(&base_url)
        .api_key(&api_key)
        .build()?;

    let collection = "client_chat_advanced_rust";

    // Insert sample data
    println!("=== Inserting Sample Data ===");
    let mut record = ekodb_client::Record::new();
    record.insert("name", "ekoDB");
    record.insert("description", "High-performance database product");
    record.insert("price", 99);
    client.insert(collection, record).await?;
    println!("✓ Inserted sample product\n");

    // Create a chat session
    println!("=== Creating Chat Session ===");
    let session = client
        .create_chat_session(
            ekodb_client::CreateChatSessionRequest::new("openai")
                .model("gpt-4")
                .collection(ekodb_client::CollectionConfig {
                    collection_name: collection.to_string(),
                    fields: vec![],
                    search_options: None,
                })
                .system_prompt("You are a helpful product assistant."),
        )
        .await?;
    let chat_id = session.chat_id.clone();
    println!("✓ Created session: {}\n", chat_id);

    // Send initial message
    println!("=== Sending Initial Message ===");
    let msg1 = client
        .chat_message(
            &chat_id,
            ekodb_client::ChatMessageRequest::new("What products are available?"),
        )
        .await?;
    println!("✓ Message sent");
    if !msg1.responses.is_empty() {
        println!("  Response: {}\n", msg1.responses[0]);
    }

    // Send a second message to have more context
    client
        .chat_message(
            &chat_id,
            ekodb_client::ChatMessageRequest::new("What is the price of ekoDB?"),
        )
        .await?;
    println!("✓ Second message sent");

    // Get the messages to find the user and assistant message IDs
    let messages = client
        .get_chat_session_messages(
            &chat_id,
            ekodb_client::GetMessagesQuery {
                limit: Some(10),
                skip: None,
                sort: Some("desc".to_string()), // Get newest first
            },
        )
        .await?;

    // Find the most recent assistant and user messages
    let assistant_message_id = messages
        .messages
        .iter()
        .find(|msg| {
            // Role is wrapped in an Object with type/value structure
            if let Some(ekodb_client::FieldType::Object(role_obj)) = msg.get("role") {
                if let Some(ekodb_client::FieldType::String(role)) = role_obj.get("value") {
                    return role == "assistant";
                }
            }
            false
        })
        .and_then(|msg| {
            // ID is a direct String field
            if let Some(ekodb_client::FieldType::String(id)) = msg.get("id") {
                Some(id.clone())
            } else {
                None
            }
        })
        .expect("Could not find assistant message");

    let user_message_id = messages
        .messages
        .iter()
        .find(|msg| {
            if let Some(ekodb_client::FieldType::Object(role_obj)) = msg.get("role") {
                if let Some(ekodb_client::FieldType::String(role)) = role_obj.get("value") {
                    return role == "user";
                }
            }
            false
        })
        .and_then(|msg| {
            if let Some(ekodb_client::FieldType::String(id)) = msg.get("id") {
                Some(id.clone())
            } else {
                None
            }
        })
        .expect("Could not find user message");

    // Feature 1: Regenerate Message
    println!("=== Feature 1: Regenerate AI Response ===");
    let regenerated = client
        .regenerate_chat_message(&chat_id, &assistant_message_id)
        .await?;
    println!("✓ Message regenerated");
    if !regenerated.responses.is_empty() {
        println!("  New response: {}\n", regenerated.responses[0]);
    }

    // Feature 2: Update Message Content
    println!("=== Feature 2: Edit Message ===");
    client
        .update_chat_message(
            &chat_id,
            &user_message_id,
            UpdateMessageRequest {
                content: "What is the price of ekoDB?".to_string(),
            },
        )
        .await?;
    println!("✓ Message content updated\n");

    // Feature 3: Toggle Forgotten Status
    println!("=== Feature 3: Mark Message as Forgotten ===");
    client
        .toggle_forgotten_message(
            &chat_id,
            &user_message_id,
            ToggleForgottenRequest { forgotten: true },
        )
        .await?;
    println!("✓ Message marked as forgotten (excluded from LLM context)\n");

    // Unmark as forgotten
    client
        .toggle_forgotten_message(
            &chat_id,
            &user_message_id,
            ToggleForgottenRequest { forgotten: false },
        )
        .await?;
    println!("✓ Message unmarked as forgotten\n");

    // Feature 4: Session Merging
    println!("=== Feature 4: Merge Chat Sessions ===");

    // Create a second session
    let session2 = client
        .create_chat_session(
            ekodb_client::CreateChatSessionRequest::new("openai")
                .model("gpt-4")
                .collection(ekodb_client::CollectionConfig {
                    collection_name: "products".to_string(),
                    fields: vec![],
                    search_options: None,
                }),
        )
        .await?;
    let chat_id2 = session2.chat_id.clone();
    println!("✓ Created second session: {}", chat_id2);

    // Send a message in the second session
    client
        .chat_message(
            &chat_id2,
            ekodb_client::ChatMessageRequest::new("Tell me more about the features"),
        )
        .await?;
    println!("✓ Sent message in second session");

    // Merge sessions
    let merge_result = client
        .merge_chat_sessions(MergeSessionsRequest {
            source_chat_ids: vec![chat_id2.clone()],
            target_chat_id: chat_id.clone(),
            merge_strategy: MergeStrategy::Chronological,
        })
        .await?;
    println!("✓ Sessions merged successfully");
    println!(
        "  Total messages in merged session: {}\n",
        merge_result.message_count
    );

    // Feature 5: Delete Message
    println!("=== Feature 5: Delete Message ===");
    client
        .delete_chat_message(&chat_id, &user_message_id)
        .await?;
    println!("✓ Message deleted\n");

    // Verify message count after deletion
    let session_details = client.get_chat_session(&chat_id).await?;
    println!("✓ Messages remaining: {}\n", session_details.message_count);

    // Cleanup
    println!("=== Cleanup ===");
    client.delete_chat_session(&chat_id).await?;
    println!("✓ Deleted session");
    client.delete_collection(collection).await?;
    println!("✓ Deleted collection\n");

    println!("✓ All advanced chat features demonstrated successfully!");

    Ok(())
}

File: examples/kotlin/examples/ClientChatAdvanced.kt
package io.ekodb.client.examples

import io.ekodb.client.EkoDBClient
import io.ekodb.client.types.Record
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import kotlinx.serialization.json.putJsonObject

/**
 * Advanced Chat example - Message regeneration, editing, and advanced features
 */
fun main() = runBlocking {
    val baseUrl = System.getenv("API_BASE_URL") ?: "http://localhost:8080"
    val apiKey = System.getenv("API_BASE_KEY") ?: error("API_BASE_KEY environment variable not set")
    
    val client = EkoDBClient.builder()
        .baseUrl(baseUrl)
        .apiKey(apiKey)
        .build()
    
    val collection = "kotlin_chat_advanced_example"
    
    println("=== ekoDB Kotlin Client - Advanced Chat Features Example ===\n")
    
    try {
        // Insert sample data
        println("=== Inserting Sample Data ===")
        
        val record = Record.new()
            .insert("name", "ekoDB")
            .insert("description", "High-performance database product")
            .insert("price", 99)
        client.insert(collection, record)
        println("✓ Inserted sample product\n")
        
        // Create a chat session
        println("=== Creating Chat Session ===")
        
        val sessionRequest = buildJsonObject {
            putJsonArray("collections") {
                add(buildJsonObject {
                    put("collection_name", collection)
                    putJsonArray("fields") {}
                })
            }
            put("llm_provider", "openai")
            put("llm_model", "gpt-4")
            put("system_prompt", "You are a helpful product assistant.")
        }
        
        val session = client.createChatSession(sessionRequest)
        val chatId = session["chat_id"]?.jsonPrimitive?.content ?: error("No chat_id in response")
        println("✓ Created session: $chatId\n")
        
        // Send initial message
        println("=== Sending Initial Message ===")
        
        val msg1Request = buildJsonObject {
            put("message", "What products are available?")
        }
        val msg1 = client.chatMessage(chatId, msg1Request)
        println("✓ Message sent")
        println("  Responses: ${msg1["responses"]}\n")
        
        // Send second message
        val msg2Request = buildJsonObject {
            put("message", "What is the price of ekoDB?")
        }
        client.chatMessage(chatId, msg2Request)
        println("✓ Second message sent\n")
        
        // Get message history to find assistant and user message IDs
        val messagesResponse = client.getChatSessionMessages(chatId, buildJsonObject {
            put("limit", 10)
            put("sort", "desc")
        })
        val messages = messagesResponse["messages"]?.jsonArray ?: error("No messages in response")
        
        // Find the most recent assistant message ID
        val assistantMessageId = messages.firstOrNull { msg ->
            val role = msg.jsonObject["role"]?.jsonObject?.get("value")?.jsonPrimitive?.content
            role == "assistant"
        }?.jsonObject?.get("id")?.jsonPrimitive?.content ?: error("No assistant message found")
        
        // Find the most recent user message ID
        val userMessageId = messages.firstOrNull { msg ->
            val role = msg.jsonObject["role"]?.jsonObject?.get("value")?.jsonPrimitive?.content
            role == "user"
        }?.jsonObject?.get("id")?.jsonPrimitive?.content ?: error("No user message found")
        
        // Regenerate the assistant message (AI response)
        println("=== Regenerating AI Response ===")
        
        val regenerated = client.regenerateChatMessage(chatId, assistantMessageId)
        println("✓ AI response regenerated")
        println("  New responses: ${regenerated["responses"]}\n")
        
        // Update a message
        println("=== Updating Message ===")
        
        val updateRequest = buildJsonObject {
            put("content", "Can you tell me more about ekoDB features?")
        }
        client.updateChatMessage(chatId, userMessageId, updateRequest)
        println("✓ Updated message content\n")
        
        // Toggle forgotten status
        println("=== Toggling Forgotten Status ===")
        
        try {
            val forgottenRequest = buildJsonObject {
                put("forgotten", true)
            }
            client.toggleForgottenMessage(chatId, userMessageId, forgottenRequest)
            println("✓ Marked message as forgotten (excluded from context)\n")
        } catch (e: Exception) {
            println("⚠ Toggle forgotten failed (connection error): ${e.message}\n")
        }
        
        // Create second session for merging
        println("=== Creating Second Session for Merge ===")
        
        val session2 = client.createChatSession(sessionRequest)
        val chatId2 = session2["chat_id"]?.jsonPrimitive?.content ?: error("No chat_id in response")
        println("✓ Created second session: $chatId2\n")
        
        // Merge sessions
        println("=== Merging Sessions ===")
        
        val mergeRequest = buildJsonObject {
            putJsonArray("source_chat_ids") {
                add(kotlinx.serialization.json.JsonPrimitive(chatId2))
            }
            put("target_chat_id", chatId)
            put("merge_strategy", "Chronological")
        }
        
        try {
            val merged = client.mergeChatSessions(mergeRequest)
            println("✓ Merged sessions")
            println("  Merged session ID: ${merged["chat_id"]}\n")
        } catch (e: Exception) {
            println("⚠ Merge failed (may not be supported): ${e.message}\n")
        }
        
        // Delete a specific message
        println("=== Deleting Message ===")
        
        client.deleteChatMessage(chatId, userMessageId)
        println("✓ Deleted message\n")
        
        // Cleanup
        println("=== Cleanup ===")
        client.deleteChatSession(chatId)
        client.deleteChatSession(chatId2)
        println("✓ Deleted chat sessions")
        
    } finally {
        try {
            client.deleteCollection(collection)
            println("✓ Deleted collection: $collection")
        } catch (e: Exception) {
            println("⚠ Could not delete collection: ${e.message}")
        }
        
        client.close()
        println("\n✓ Advanced chat features example completed successfully")
    }
}
